### ISSUE TITLE
Application for KPFM

### ISSUE THREAD

**Issue by Aimannor** [Probe-Particle/ppafm#133]
Can the model be used for modeling KPFM for charged surfaces?

**Comment by aureliojgc** [comment]
Hi,
The link you had send Niko is not the version of KPFM that I implemented. About the information of my version. I think I did not add it to the wiki yet, my bad. I will prepare something ASAP and add it.

**Comment by aureliojgc** [comment]
about the charged surfaces... As long as you can calculate the charged state with DFT and you assume that the extra charge in the surface will not induce any charge transference between the tip and sample and that the tip will not induce any charge rearrangement on the surface it should be fine...

**Comment by Aimannor** [comment]
Thank you for getting back to me so quickly. I have more related questions to ask.
1) Can the model be used for AM or FM KPFM modeling?
2) The other thing is, if I have 2D surface potential and topography contour can I convert them to charge state contour for unknown material using the model?

**Comment by aureliojgc** [comment]
Hi! I am sorry I missed the last messages. I hope you didn't need the answer much in a hurry.

About the modulation, the AFM code only works on FM and since the KPFM is an extension of it, it will only work on FM. maybe @ProkopHapala can spreed further light on this.

and for the surface potential... we never tried something like it as long as I am aware... but if you find the way to extend your data to an electrostatic potential in a tridimensional cube of xsf it will maybe go through. but will be unknown territoty as long I know.

Sorry again for the late answer,
A.

**Comment by mondracek** [comment]
**The KPFM functionality is not working!**
I am attaching an example in which I have attempted to calculate the LCPD on 2 atomic layers of NaCl on the Au(001). The attached archive file contains the inputs (``params.ini``, ``hatree_potential.cube`` with the electrostatic potential of the sample, and ``hartree_potential_Vz.cube`` with the electrostatic potential of the sample in an external field), the script ``run.sh`` with commands to run and the output file ``stdout+err.txt`` that contains the standard output and standard error output messages.
[NaCl-on-Au001.tar.gz](https://github.com/Probe-Particle/ppafm/files/12231656/NaCl-on-Au001.tar.gz)
What I figure is happening: The command
```ppafm-relaxed-scan --Vrange -0.5 0.5 3```
should write the forces into
```Q-0.20K0.25V-0.50/OutFz.xsf```,
```Q-0.20K0.25V0.00/OutFz.xsf``` and
```Q-0.20K0.25V0.50/OutFz.xsf```,
according to the bias (_V_), where the subsequent command
``ppafm-plot-results --Vrange -0.5 0.5 3 --LCPD_maps --cbar --V0 0 --WSxM``
expects to find them. But in fact, ``ppafm-relaxed-scan`` only generates
```Q-0.20K0.25/OutFz.xsf```,
probably overwriting results for one _V_ with another. Consequently, ``ppafm-plot-results`` complains about not being able to load the files it needs. This error does not immediately terminate the execution (perhaps it should) but as a result of the unavailable data, ``UnboundLocalError: local variable 'LCPD_b' referenced before assignment`` happens, upon which the script finally fails.
By the way, without the option ``--WSxM``, which asks for an output in the (proprietary?) WSxM format (essentially text data files), ``plot_results.py`` does not even attempt to open the ``Q-0.20K0.25V*/OutFz.xsf`` files (and then it dies by unassigned ``LCPD_b`` nonetheless). This seems wrong too: The ``--LCPD_maps`` option, which asks for LCPD in the PNG format, should be enough to make the program look for the ``Q*K*V*/OutFz.xsf`` files.

**Comment by NikoOinonen** [comment]
@mondracek Since you mentioned the `Vrange` which I modified in my recent PR, I looked into it. It looks to me like the incorrect folders are caused by this line in `relaxed_scan.py`: https://github.com/Probe-Particle/ppafm/blob/6cad5262307cffecc96a3b994cb6f95d93ac10b6/ppafm/cli/relaxed_scan.py#L129
It overwrites the directory name even when the bias is applied. I tried removing this line and the example you posted ran to completion. However, the produced images were empty, so there might be something else going wrong as well. I don't know the KPFM code, so I can't investigate any further.
